lib/core/src/Cardano/Wallet/Api/Server.hs

```diff
@@ -674,6 +677,101 @@ postShelleyWallet ctx body = do
+                    (ApiDelegationStatus $
+                        ApiT $ W.Delegating (ApiT poolId))
+                    Nothing
+    [ (W.DelegationDiscovered (W.SlotId epochN1 _) poolIdMaybe1)
+        , (W.DelegationDiscovered (W.SlotId epochN2 _) poolIdMaybe2)] ->
+        if epochN2 > epochN1 then
+            error "delegation discovered later cannot have epoch number \
+                  \older than the older discovered delegation"
+        else case (poolIdMaybe1, poolIdMaybe2) of
+            (Nothing, Nothing) ->
+                error "there should not be two quit delegation in row"
+            (Just poolId, Nothing) ->
+                if currentEpochNo < epochN1 + 2 then
+                    ApiWalletDelegation
+                    (ApiDelegationStatus $ ApiT W.NotDelegating)
+                    (Just $
+                        ApiWalletDelegationNext
+                        (ApiDelegationStatus $ ApiT $
+                            W.Delegating (ApiT poolId))
+                        (ApiEpochInfo
+                            (ApiT $ epochN1 + 2)
+                            (W.epochStartTime sp (epochN1 + 2)))
+                    )
+                else
+                    ApiWalletDelegation
+                    (ApiDelegationStatus $ ApiT $
+                        W.Delegating (ApiT poolId))
+                    Nothing
+            (Just poolId1, Just poolId2) ->
+                if currentEpochNo < epochN1 + 2 then
+                    ApiWalletDelegation
+                    (ApiDelegationStatus $ ApiT $
+                        W.Delegating (ApiT poolId2))
+                    (Just $
+                        ApiWalletDelegationNext
+                        (ApiDelegationStatus $ ApiT $
+                            W.Delegating (ApiT poolId1))
+                        (ApiEpochInfo
+                            (ApiT $ epochN1 + 2)
+                            (W.epochStartTime sp (epochN1 + 2)))
+                    )
+                else
+                    ApiWalletDelegation
+                    (ApiDelegationStatus $ ApiT $
+                        W.Delegating (ApiT poolId1))
+                    Nothing
+            (Nothing, Just poolId) ->
+                if currentEpochNo < epochN1 + 2 then
+                    ApiWalletDelegation
+                    (ApiDelegationStatus $ ApiT $
+                        W.Delegating (ApiT poolId))
+                    (Just $
+                        ApiWalletDelegationNext
+                        (ApiDelegationStatus $ ApiT W.NotDelegating)
+                        (ApiEpochInfo
+                            (ApiT $ epochN1 + 2)
+                            (W.epochStartTime sp (epochN1 + 2)))
+                    )
+                else
+                    ApiWalletDelegation
+                    (ApiDelegationStatus $ ApiT W.NotDelegating)
+                    Nothing
+    _ ->
+        error "takeDelegationsDiscovered cannot return more than two \
```

Comment:
maybe, but I think it is overstretch here. We use DBLayer function that suppose to return **at most** two elements (we have `take 2`), so I have this error here although I think it will "never" happen